src/webapi/indexeddb.rs

```diff
@@ -0,0 +1,1145 @@
+                return @{self.as_ref()}.put(@{value.as_ref()}, @{key.as_ref()});
+            )
+        }.unwrap()
+    }
+    
+    /// Returns an `IDBRequest` object, and, in a separate thread, creates a structured clone of the value, and stores the cloned value in the object store. This is for adding new records to an object store.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/add)
+    pub fn add<T: Into<Option<Value>>>( &self, value: Value, key: T) -> Result<IDBRequest, IDBAddError> {
+        match key.into() {
+            None => js_try! (
+                return @{self.as_ref()}.add(@{value.as_ref()});
+            ),
+            Some(key) => js_try! (
+                return @{self.as_ref()}.add(@{value.as_ref()}, @{key.as_ref()});
+            )
+        }.unwrap()
+    }
+
+    /// returns an `IDBRequest` object, and, in a separate thread, deletes the specified record or records.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/delete)
+    pub fn delete( &self, query: Value) -> Result<IDBRequest, IDBObjectStoreDeleteError> {
+        js_try! (
+            return @{self.as_ref()}.delete(@{query.as_ref()});
+        ).unwrap()
+    }
+    
+    /// Returns an IDBRequest object, and clears this object store in a separate thread
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/clear)
+    pub fn clear( &self ) -> Result<IDBRequest, IDBClearError> {
+        js_try! (
+            return @{self.as_ref()}.clear();
+        ).unwrap()
+    }
+    
+    /// opens a named index in the current object store
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/index)
+    pub fn index( &self, name: &str) -> Result<IDBIndex, IDBIndexError> {
+        js_try! (
+            return @{self.as_ref()}.index(@{name});
+        ).unwrap()
+    }
+
+    // [NewObject] IDBIndex createIndex(DOMString name, (DOMString or sequence<DOMString>) keyPath, optional IDBIndexParameters options);
+    /// Creates and returns a new `IDBIndex` object in the connected database.
+    ///
+    /// Note that this method must be called only from a VersionChange
+    /// transaction mode callback.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/createIndex)
+    pub fn create_index( &self, name: &str, key_path: &str, options: Value) -> IDBIndex { // TODO, how am I doing the optinal options?
+        js! (
+            return @{self.as_ref()}.createIndex(@{name}, @{key_path}, @{options.as_ref()});
+        ).try_into().unwrap()
+    }
+    
+    /// Destroys the index with the specified name in the connected database, used during a version upgrade.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/deleteIndex)
+    pub fn delete_index( &self, name: &str) {
```

Comment:
This can throw according to the specs so should return `Result`.

Reply:
Trying to fix this, should the return value be `Result<(), IDBIndexError>` or ?